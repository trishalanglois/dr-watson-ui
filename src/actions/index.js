export const createUser = user => ({
  type: 'CREATE_USER',
  user
});

export const removeUser = () => ({
  type: 'REMOVE_USER'
});

export const hasErrored = errorMsg => ({
  type: 'HAS_ERRORED',
  errorMsg
})

export const addMessage = message => ({
  type: 'ADD_MESSAGE',
  message
})

export const clearMessages = () => ({
  type: 'CLEAR_MESSAGES',
  messages: []
})
