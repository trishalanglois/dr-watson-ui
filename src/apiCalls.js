export const startConversation = async feeling => {
  const url = 'https://drwatson-api.herokuapp.com/api/v1/start_session'
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ feeling })
  };

  const response = await fetch(url, options);
  if (!response.ok) {
    throw Error('Dr Watson is currently down.  Please try again later.')
  }

  const data = await response.json();
  console.log(data)
  return data;
}

export const postMessage = async newMessage => {
  const url = 'https://drwatson-api.herokuapp.com/api/message'
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ newMessage })
  };

  const response = await fetch(url, options);
  if (!response.ok) {
    throw Error('Dr Watson is currently down.  Please try again later.')
  }

  const data = await response.json();
  console.log(data);
  return data;
}

export const endConversation = async () => {
  const url = 'https://drwatson-api.herokuapp.com/api/v1/end_session';

  const response = await fetch(url);
  if (!response.ok) {
    throw Error('There was a problem ending the session.  Please close the application.')
  }
}
