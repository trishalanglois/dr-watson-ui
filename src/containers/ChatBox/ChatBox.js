import React, { Component, createRef } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { hasErrored } from '../../actions';
import { postMessage } from '../../apiCalls';
import Message from '../../components/Message/Message';
import { addMessage } from '../../actions';

import "./ChatBox.css"

export class ChatBox extends Component {
  constructor() {
    super();
    this.state = { message: '' }
    this.convo = createRef();
  }

  componentDidUpdate() {
    this.convo.scrollTop = this.convo.scrollHeight;
  }

  handleChange = e => {
    this.setState({ message: e.target.value });
  }

  handleSubmit = e => {
    if (e.key === 'Enter' || e.button === 0) {
      const { message } = this.state;
      this.props.addMessage(message, true);
      this.setState({ message: '' });
      this.messageChatBot();
    }
  }

  messageChatBot = async () => {
    try {
      const messageResponse = await postMessage(this.state.message);
      this.props.addMessage(messageResponse.message, false);
    } catch({ message }) {
      this.props.hasErrored(message)
    }
  }

  render() {
    const { message } = this.state;
    const { messages, errorMsg } = this.props;
    const survey = messages.map((message, i) => {
      return <Message
        key={`message${i}`}
        message={message}
      />
    })
    return (
      <main className="chat-container">
        <section className="conversation" ref={node => this.convo = node}>
          {survey}
          {errorMsg && <p className="message watson error">{errorMsg}</p>}
        </section>
        <section className="messenger">
          <input
            placeholder='Chat with Survey Bot here...'
            value={message}
            onChange={this.handleChange}
            onKeyPress={this.handleSubmit}
          />
          <button onClick={this.handleSubmit}>Submit</button>
        </section>
      </main>
    )
  }
}

export const mapStateToProps = ({ errorMsg, messages }) => ({
  errorMsg,
  messages
})

export const mapDispatchToProps = dispatch => bindActionCreators({ hasErrored, addMessage }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ChatBox);
