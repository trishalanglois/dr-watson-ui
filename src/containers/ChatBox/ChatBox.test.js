import React from 'react';
import { shallow, mount } from 'enzyme';
import { ChatBox, mapStateToProps, mapDispatchToProps } from './ChatBox';
import { hasErrored, addMessage } from '../../actions';
import { postMessage } from '../../apiCalls';


jest.mock('../../apiCalls');

describe('ChatBox component', () => {
  let wrapper
  const mockHasErrored = jest.fn();
  const mockAddMessage = jest.fn();
  const mockMessages = [
    {
      message: "Hello, I am Dr. Watson.  My understanding is that you are feeling tired.  Have you been feeling anxious this week?",
      i: 0,
      isUser: false
    },
    {
      message: "Absolutely.  I feel like I am struggling to keep up.",
      i: 1,
      isUser: true
    }
  ];

  beforeEach(() => {
    wrapper = shallow(<ChatBox
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);
  });

  it('should match the snapshot rendering all messages', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot rendering an error if there is an errorMsg in the store', () => {
    wrapper = shallow(<ChatBox
      messages={mockMessages}
      errorMsg={'fetch failed.'}
      hasErrored={mockHasErrored}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should update the state when handleChange is called', () => {
    wrapper = mount(<ChatBox
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);

    wrapper.instance().handleChange({ target: { value: 'Hello world' } });

    expect(wrapper.state('message')).toEqual('Hello world');
  });

  it('should call messageChatBot, and clear state when calling handleSubmit pressing Enter', () => {
    wrapper = mount(<ChatBox
      addMessage={mockAddMessage}
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);
    wrapper.instance().messageChatBot = jest.fn();

    wrapper.setState({ message: 'Hello world' });
    wrapper.instance().handleSubmit({ key: 'Enter' });

    expect(wrapper.state('message')).toEqual('');
    expect(wrapper.instance().messageChatBot).toHaveBeenCalled();
  });

  it('should call addMessage, messageChatBot, and clear state when calling handleSubmit clicking the button', () => {
    wrapper = mount(<ChatBox
      addMessage={mockAddMessage}
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);
    wrapper.instance().messageChatBot = jest.fn();

    wrapper.setState({ message: 'Hello world' });
    wrapper.instance().handleSubmit({ button: 0 });

    expect(wrapper.state('message')).toEqual('');
    expect(wrapper.instance().messageChatBot).toHaveBeenCalled();
  });

  it('should call postMessage and addMessage when calling messageChatBot', async () => {
    wrapper = mount(<ChatBox
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);

    postMessage.mockImplementation(() => {
      return Promise.resolve({ message: 'My name is Dr. Watson.  How are you today?' });
    });

    wrapper.instance().setState({ message: 'Hi there.' });
    await wrapper.instance().messageChatBot();

    expect(postMessage).toHaveBeenCalledWith('Hi there.')
  });

  it('should call hasErrored if messageChatBot rejects', async () => {
    wrapper = mount(<ChatBox
      messages={mockMessages}
      hasErrored={mockHasErrored}
    />);

    postMessage.mockImplementation(() => {
      return Promise.reject(Error('fetch failed.'));
    });

    await wrapper.instance().messageChatBot();

    expect(mockHasErrored).toHaveBeenCalledWith('fetch failed.')
  });
});

describe('mapStateToProps', () => {
  it('should return an object with the messages and errorMsg information', () => {
    const mockUser = {
      id: 1568665187737,
      firstName: "Travis",
      lastName: "Rollins",
      feeling: "tired"
    };

    const mockState = {
      user: mockUser,
      messages: [{
        message: 'Hi there, my name is Dr. Watson. I understand that you have been feeling happy. That is super exciting to hear!',
        isUser: false,
      }],
      errorMsg: ''
    };
    const expected = {
      messages: [{
        message: 'Hi there, my name is Dr. Watson. I understand that you have been feeling happy. That is super exciting to hear!',
        isUser: false,
      }],
      errorMsg: ''
    };
    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  it('calls dispatch with a hasErrored action when hasErrored is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = hasErrored('fetch failed');

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.hasErrored('fetch failed');

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with a addMessage action when addMessage is called', () => {
    const mockMessage = {
      message: 'Eric is the best 1908 dad',
      id: 3
    };
    const mockDispatch = jest.fn();
    const actionToDispatch = addMessage(mockMessage);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addMessage(mockMessage);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

  })
});
