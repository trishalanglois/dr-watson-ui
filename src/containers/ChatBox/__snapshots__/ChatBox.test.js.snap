// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBox component should match the snapshot rendering all messages 1`] = `
<main
  className="chat-container"
>
  <section
    className="conversation"
  >
    <Message
      key="message0"
      message="Hello, I am Dr. Watson.  My understanding is that you are feeling tired.  Have you been feeling anxious this week?"
    />
    <Message
      key="message1"
      message="Absolutely.  I feel like I am struggling to keep up."
    />
  </section>
  <section
    className="messenger"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Chat with Survey Bot here..."
      value=""
    />
    <button
      onClick={[Function]}
    >
      Submit
    </button>
  </section>
</main>
`;

exports[`ChatBox component should match the snapshot rendering an error if there is an errorMsg in the store 1`] = `
<main
  className="chat-container"
>
  <section
    className="conversation"
  >
    <Message
      key="message0"
      message="Hello, I am Dr. Watson.  My understanding is that you are feeling tired.  Have you been feeling anxious this week?"
    />
    <Message
      key="message1"
      message="Absolutely.  I feel like I am struggling to keep up."
    />
    <p
      className="message watson error"
    >
      fetch failed.
    </p>
  </section>
  <section
    className="messenger"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Chat with Survey Bot here..."
      value=""
    />
    <button
      onClick={[Function]}
    >
      Submit
    </button>
  </section>
</main>
`;
